### PR TITLE
Mixed double precision for PPO algorithm 

### PR DESCRIPTION
**Pull request by lopatovsky**
# Mixed precision<a id="mixed-precision"></a>

**Motivation**:

Inspired by RLGames, we implemented automatic mixed double precision to boost performance of PPO.

**Sources:**

<https://pytorch.org/docs/stable/amp.html>

<https://pytorch.org/docs/stable/notes/amp_examples.html>

**Speed eval:**

- Big neural network (units: \[2048, 1024, 1024, 512])

- 10000 steps

- Running on top of Oige env simulation (constant for each run)

- Skrl uses single forward pass implementation

|             |                     |              |                                                   |
| :---------: | :-----------------: | :----------: | :-----------------------------------------------: |
| **Library** | **Mixed-Precision** | **Time (s)** | **slowing factor** Base: rlgames, mixed pr. = True |
|   RLGames   |          No         |      448     |                       1.322x                      |
|   RLGames   |         Yes         |      339     |                      1 (base)                     |
|     SKRL    |          No         |      475     |                       1.401x                      |
|     SKRL    |         Yes         |      373     |                        1.1x                       |
|     SKRL    |        Yes \*       |      358     |                       1.056x                      |

\* in this run mixed precision was used also for inference during data collection phase

**Quality eval:**

- We trained a policy for our task with each of the configurations multiple times. We didn’t observe any statistically significant difference in quality of the final results.

### CODE REVIEW

File: skrl/agents/torch/ppo/ppo.py [hunk 1]
```diff
@@ -388,55 +398,62 @@ def compute_gae(rewards: torch.Tensor,
             # mini-batches loop
             for sampled_states, sampled_actions, sampled_log_prob, sampled_values, sampled_returns, sampled_advantages in sampled_batches:
 
-                sampled_states = self._state_preprocessor(sampled_states, train=not epoch)
+                with torch.autocast(device_type=self._device_type, enabled=self._mixed_precision):
```
Comment (Toni-SM) [inline skrl/agents/torch/ppo/ppo.py:401]: Is it necessary to apply `autocast` to https://github.com/Toni-SM/skrl/blob/c15f3ce6f6d614d2e433439f9ed0606826a5e45a/skrl/agents/torch/ppo/ppo.py#L369-L373?

File: skrl/agents/torch/ppo/ppo.py [hunk 2]
```diff
@@ -219,8 +227,9 @@ def act(self, states: torch.Tensor, timestep: int, timesteps: int) -> torch.Tens
             return self.policy.random_act({"states": self._state_preprocessor(states)}, role="policy")
 
         # sample stochastic actions
-        actions, log_prob, outputs = self.policy.act({"states": self._state_preprocessor(states)}, role="policy")
-        self._current_log_prob = log_prob
+        with torch.autocast(device_type=self._device_type, enabled=(self._mixed_precision)):
```
Comment (Toni-SM) [inline skrl/agents/torch/ppo/ppo.py:230]: Why `self._mixed_precision` in all `enabled=(self._mixed_precision)` statements is between `()`?

Comment (lopatovsky) [reply skrl/agents/torch/ppo/ppo.py:230]: it's unnecessery (legacy from testing). removed.

File: skrl/agents/torch/ppo/ppo.py (hunk 1 above)
Comment (lopatovsky) [reply skrl/agents/torch/ppo/ppo.py:401]: Autocasting automatically chooses the precision for GPU operations to improve performance while maintaining accuracy [doc](https://pytorch.org/docs/stable/notes/amp_examples.html) 

removing the autocast will decrease the performance because float16 will not be used where it can be used.
I ran a small experiment:
| Scenario                                  | Time  |
|-------------------------------------------|-------|
| Autocast block as it is                   | 1:46  |
| No autocast                               | 2:10  |
| Autocast block only from "# compute value loss" | 1:50  |
| Autocast block only from "# compute entropy loss" | 1:49  |
